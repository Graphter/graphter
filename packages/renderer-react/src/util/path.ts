import { PathSegment } from "@graphter/core";


export function validate(path?: Array<PathSegment> | null){
  if(typeof path === null || path === undefined) return {
    valid: false,
    reason: 'Path is empty'
  }
  if(!Array.isArray(path)) return {
    valid: false,
    reason: 'Path is not an array'
  }
  for(const segment of path) {
    switch (typeof segment) {
      case 'string':
        break
      case 'number':
        break
      default:
        return {
          valid: false,
          reason: `'${typeof segment}' is an invalid path segment`
        }
    }
  }
  return {
    valid: true
  }
}

export const getValue = (data: any, path?: Array<PathSegment> | null, defaultVal?: any) => {
  const targetData = path?.reduce((data: any, pathSegment: PathSegment) => {
    if(typeof data !== 'undefined') return data[pathSegment]
  }, data)
  if(typeof targetData === 'undefined'){
    if(typeof defaultVal === 'undefined')
      throw new Error(`Could not find a value at ${path?.join('/')} within: ${JSON.stringify(data)} and no default value was given.`)
    else return defaultVal
  }
  return targetData
}

export const pathUtils = {
  validate,
  getValue
}