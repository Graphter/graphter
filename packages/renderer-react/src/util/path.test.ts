import { getValue, validate } from "./path";

describe('path', () => {
  describe('validate()', () => {
    it.each([
      {path: [ ]},
      {path: [ 'page' ]},
      {path: [ 'page', 0 ]},
      {path: [ 'page', 0, 'title' ]}
    ])
    ('should return valid for valid path (%o)', (validPath) => {
      const result = validate(validPath.path)
      expect(result.valid).toBe(true)
    })
    it.each([ null, undefined ])
    ('should return invalid if the path is %s', (emptyPath) => {
      const result = validate(emptyPath)
      expect(result.valid).toBe(false)
      expect(result.reason).toMatchSnapshot()
    })
    it.each([ 'path', 0, 2, {} ])
    ('should return invalid if the path is not an array (%s)', (nonArrayPath) => {
      // @ts-ignore
      const result = validate(nonArrayPath)
      expect(result.valid).toBe(false)
      expect(result.reason).toBe('Path is not an array')
    })
    it.each([
      null,
      undefined,
      {},
      []
    ])('should return invalid if any of the segments are of the wrong type (%s)', (invalidSegment) => {
      // @ts-ignore
      const result = validate([ 'page', invalidSegment, 0 ])
      expect(result.valid).toBe(false)
      expect(result.reason).toMatchSnapshot()
    })
  })
  describe('getValue()', () => {
    it.each([
      [ [], 'the-value' ],
      [ [ 'foo' ], { foo: 'the-value' } ],
      [ [ 'foo', 'bar' ], { foo: { bar: 'the-value'} } ],
      [ [ 'foo', 0 ], { foo: [ 'the-value' ] } ],
      [ [ 'foo', 0, 'bar' ], { foo: [ { bar: 'the-value'} ] } ],
      [ [ 'foo', 1, 'baz' ], { foo: [ { bar: 'some-other-value'}, { baz: 'the-value'} ] } ],
    ])
    ('should return the correct value when looking for %o in %o', (path, data) => {
      const result = getValue(data, path)
      expect(result).toBe('the-value')
    })
    describe.each([
      [ [ 'foo', 0, 'bar' ], 'the-value' ],
      [ [ 'foo', 0, 'bar' ], { foo: 'the-value' } ],
      [ [ 'foo', 0, 'bar' ], { foo: [ 'the-value' ] } ],
    ])('when data cannot be found at path %o in %o', (path, data) => {
      it('should return the default value if supplied', () => {
        const result = getValue(data, path, 'the-default-value')
        expect(result).toBe('the-default-value')
      })
      it('should error if default value is not defined', () => {
        expect.assertions(1)
        try {
          getValue(data, path)
        } catch(err) {``
          expect(err).toMatchSnapshot()
        }
      })
    })
    it.each([null, undefined, {}, []])
    ('should error if it comes across bad path segment %o', (badSegment) => {
      expect.assertions(1)
      try {
        // @ts-ignore
        getValue({ some: { nested: 'data' } }, [ 'some', badSegment ])
      } catch(err) {
        expect(err).toMatchSnapshot()
      }
    })
    it('should not mutate the path', () => {
      const path = [ 'some' ]
      Object.freeze(path)
      getValue({ some: { nested: 'data' } }, path)
    })
  })
})

export {}