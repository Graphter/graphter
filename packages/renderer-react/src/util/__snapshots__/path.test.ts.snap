// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`path getValue() should error if it comes across bad path segment [ [length]: 0 ] 1`] = `[Error: Could not find a value at some/ within: {"some":{"nested":"data"}} and no default value was given.]`;

exports[`path getValue() should error if it comes across bad path segment {} 1`] = `[Error: Could not find a value at some/[object Object] within: {"some":{"nested":"data"}} and no default value was given.]`;

exports[`path getValue() should error if it comes across bad path segment null 1`] = `[Error: Could not find a value at some/ within: {"some":{"nested":"data"}} and no default value was given.]`;

exports[`path getValue() should error if it comes across bad path segment undefined 1`] = `[Error: Could not find a value at some/ within: {"some":{"nested":"data"}} and no default value was given.]`;

exports[`path getValue() when data cannot be found at path [ 'foo', 0, 'bar', [length]: 3 ] in 'the-value' should error if default value is not defined 1`] = `[Error: Could not find a value at foo/0/bar within: "the-value" and no default value was given.]`;

exports[`path getValue() when data cannot be found at path [ 'foo', 0, 'bar', [length]: 3 ] in { foo: 'the-value' } should error if default value is not defined 1`] = `[Error: Could not find a value at foo/0/bar within: {"foo":"the-value"} and no default value was given.]`;

exports[`path getValue() when data cannot be found at path [ 'foo', 0, 'bar', [length]: 3 ] in { foo: [ 'the-value', [length]: 1 ] } should error if default value is not defined 1`] = `[Error: Could not find a value at foo/0/bar within: {"foo":["the-value"]} and no default value was given.]`;

exports[`path validate() should return invalid if any of the segments are of the wrong type ([]) 1`] = `"'object' is an invalid path segment"`;

exports[`path validate() should return invalid if any of the segments are of the wrong type ({}) 1`] = `"'object' is an invalid path segment"`;

exports[`path validate() should return invalid if any of the segments are of the wrong type (null) 1`] = `"'object' is an invalid path segment"`;

exports[`path validate() should return invalid if any of the segments are of the wrong type (undefined) 1`] = `"'undefined' is an invalid path segment"`;

exports[`path validate() should return invalid if the path is null 1`] = `"Path is not an array"`;

exports[`path validate() should return invalid if the path is undefined 1`] = `"Path is empty"`;
